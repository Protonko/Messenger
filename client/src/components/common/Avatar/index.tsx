import {useMemo, FC} from 'react'
import classNames from 'classnames'
import {Sizes} from 'models/common/sizes'
import {ColorGeneratorByName} from 'utils/ColorGeneratorByName'
import {isDarkColor} from 'utils/isDarkColor'
import {Loader} from 'components/common/Loader'

export interface IPropsAvatar {
  src: string
  name?: string
  size?: Sizes
  customStyles?: string
}

export const Avatar: FC<IPropsAvatar> = ({
  src,
  name,
  size = Sizes.MEDIUM,
  customStyles,
}) => {
  const backgroundColor = useMemo(
    () => new ColorGeneratorByName(name ?? '').generate(),
    [src, name],
  )
  const avatarClassname = classNames(
    'avatar',
    {'avatar--sm': size === Sizes.SMALL},
    {'avatar--lg': size === Sizes.LARGE},
    {[customStyles ?? '']: customStyles},
  )

  if (!src && name) {
    const classNamesText = classNames(
      'avatar__text',
      {'avatar__text--light': isDarkColor(backgroundColor)},
      {'avatar__text--dark': !isDarkColor(backgroundColor)},
    )

    return (
      <div className={avatarClassname} style={{backgroundColor}}>
        <span className={classNamesText}>{name.charAt(0)}</span>
      </div>
    )
  }

  return (
    <div className={avatarClassname}>
      <img className="avatar__image" src={src} alt="" />
      <Loader />
    </div>
  )
}
