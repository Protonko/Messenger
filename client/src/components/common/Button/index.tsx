import {FC, HTMLProps, ReactNode} from 'react'

interface IButtonProps extends HTMLProps<HTMLButtonElement> {
  text?: string
  icon?: ReactNode
  leftAddons?: ReactNode
  rightAddons?: ReactNode
  type?: 'button' | 'submit' | 'reset'
}

export const Button: FC<IButtonProps> = ({
  text,
  icon,
  leftAddons,
  rightAddons,
  ...buttonProps
}) => {
  const renderButtonContent = () => {
    return (
      <>
        {text && (
          <span className="button__text">
            {text}
          </span>
        )}
        {icon && (
          <span className="button__icon">
            {icon}
          </span>
        )}
      </>
    )
  }

  return (
    <button className="button" {...buttonProps}>
      {leftAddons}
      {renderButtonContent()}
      {rightAddons}
    </button>
  )
}
