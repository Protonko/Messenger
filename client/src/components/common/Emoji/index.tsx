import 'emoji-mart/css/emoji-mart.css'
import {FC, useState, useRef} from 'react'
import {EmojiData, Picker} from 'emoji-mart'
import {ReactComponent as Smile} from 'assets/icons/smile.svg'
import {useOutsideClick} from 'hooks/useOutsideClick'

export interface IPropsEmoji {
  onSelect: (emoji: EmojiData) => void
}

export const Emoji: FC<IPropsEmoji> = ({onSelect}) => {
  const [open, setOpen] = useState(false)
  const ref = useRef(null)
  const initiatorRef = useRef(null)

  useOutsideClick(ref, () => setOpen(false), initiatorRef)

  const onSelectEmoji = (emoji: EmojiData) => {
    onSelect(emoji)
    setOpen(false)
  }

  return (
    <div className="emoji">
      {open && (
        <div className="emoji__popup" ref={ref}>
          <Picker native={true} onSelect={onSelectEmoji} />
        </div>
      )}

      <Smile
        className="emoji__button"
        ref={initiatorRef}
        onClick={() => setOpen(!open)}
      />
    </div>
  )
}
