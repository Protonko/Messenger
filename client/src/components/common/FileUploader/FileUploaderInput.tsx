import type {ChangeEvent, FC} from 'react'
import {useDispatch} from 'react-redux'
import classNames from 'classnames'
import {ReactComponent as Clip} from 'assets/icons/clip.svg'
import {uploadFile} from 'store/actions/files'

interface FileUploaderProps {
  onChange?: () => void
  multiple?: boolean
  disabled?: boolean
}

export const FileUploader: FC<FileUploaderProps> = ({
  onChange,
  multiple,
  disabled,
}) => {
  const dispatch = useDispatch()
  const uploaderClassNames = classNames('file-uploader', {
    'file-uploader--disabled': disabled,
  })

  const onUpload = (event: ChangeEvent<HTMLInputElement>) => {
    onChange?.()
    console.log(event.target.files)
    dispatch(uploadFile(event.target.files))
  }

  return (
    <div className={uploaderClassNames}>
      <label className="file-uploader__label">
        <Clip className="file-uploader__icon" />
        <input
          className="file-uploader__input"
          type="file"
          onChange={onUpload}
          multiple={multiple}
          disabled={disabled}
        />
      </label>
    </div>
  )
}
