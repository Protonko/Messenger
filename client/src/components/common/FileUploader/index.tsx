import type {ChangeEvent, FC} from 'react'
import classNames from 'classnames'
import {ReactComponent as Clip} from 'assets/icons/clip.svg'

export interface IFileUploaderProps {
  onChange: (file: File) => void
  disabled?: boolean
}

export const FileUploader: FC<IFileUploaderProps> = ({
  onChange,
  disabled,
}) => {
  const uploaderClassNames = classNames('file-uploader', {
    'file-uploader--disabled': disabled,
  })

  const onUpload = (event: ChangeEvent<HTMLInputElement>) => {
    const file =  event.target.files?.[0]

    file && onChange(file)
  }

  return (
    <div className={uploaderClassNames}>
      <label className="file-uploader__label">
        <Clip className="file-uploader__icon" />
        <input
          name="file"
          className="file-uploader__input"
          type="file"
          onChange={onUpload}
          disabled={disabled}
        />
      </label>
    </div>
  )
}
