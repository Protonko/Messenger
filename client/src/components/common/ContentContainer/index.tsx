import type {FC} from 'react'
import {Loader} from 'components/common/Loader'
import {Text} from 'components/common/Text'

interface ContentContainerProps {
  loading: boolean
  errorMessage: string | null
  children: JSX.Element
}

export const ContentContainer: FC<ContentContainerProps> = ({
  loading,
  errorMessage,
  children,
}) => {
  if (errorMessage) {
    return <Text>{errorMessage}</Text>
  }

  if (loading) {
    return <Loader center={true} />
  }

  return children
}
