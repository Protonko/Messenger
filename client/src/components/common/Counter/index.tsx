// types
import {Status} from 'models/common/status'

import {FC} from 'react'
import classNames from 'classnames'

export interface IPropsCounter {
  count: number | '99+',
  status?: Status,
}

export const Counter: FC<IPropsCounter> = ({
  count,
  status = Status.ACTIVE,
}) => {
  const classNameStatus = classNames(
    'counter',
    {'counter--muted': status === Status.MUTED}
  )

  return (
    <div className={classNameStatus}>
      <span className="counter__text">
        {count}
      </span>
    </div>
  )
}
