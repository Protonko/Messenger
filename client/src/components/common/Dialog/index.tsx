import type {FC} from 'react'
import type {IDialog} from 'models/dialog'
import {ReadStatus} from 'models/common/status'
import {TextSize, TextTypes} from 'models/common/text'
import {ReactComponent as DoubleCheck} from 'assets/icons/double-check.svg'
import {ReactComponent as Check} from 'assets/icons/check.svg'
import {COLORS} from 'static/colors'
import {Text} from 'components/common/Text'
import {Counter} from 'components/common/Counter'
import {Avatar} from 'components/common/Avatar'

export const Dialog: FC<IDialog> = ({
  name,
  description,
  avatar,
  edited,
  createdAt,
  updatedAt,
  messages,
  status,
  readStatus,
}) => {
  const counter = messages.toString().length > 2 ? '99+' : messages

  const renderStatus = () => {
    if (!!counter) {
      return <Counter count={counter} status={status}/>
    }

    if (readStatus) {
      const props = {
        width: 15,
        height: 15,
        color: COLORS.dustyGray,
      }
      return readStatus === ReadStatus.READ ? <DoubleCheck {...props} /> : <Check {...props} />
    }

    return null
  }

  return (
    <div className="dialog">
      <Avatar src={avatar} />

      <div className="dialog__data">
        <div className="dialog__text">
          <Text
            type={TextTypes.h4}
            customStyles="dialog__text-title"
            numberOfLines={1}
          >
            {name}
          </Text>
          <Text>{description}</Text>
        </div>

        <div className="dialog__info">
          <Text
            type={TextTypes.mixed}
            size={TextSize.EXTRA_SMALL}
            customStyles="dialog__info-date"
          >
            {new Date(edited ? updatedAt : createdAt).toLocaleDateString()}
          </Text>

          {renderStatus()}
        </div>
      </div>
    </div>
  )
}
