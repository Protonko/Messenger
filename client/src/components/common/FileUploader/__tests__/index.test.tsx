import {mount, ReactWrapper} from 'enzyme'
import {FileUploader, IFileUploaderProps} from 'components/common/FileUploader'

const ELEMENT_SELECTORS = {
  input: '.file-uploader__input',
}

describe('FileUploader', () => {
  let component: ReactWrapper<IFileUploaderProps>
  let props: IFileUploaderProps

  beforeEach(() => {
    props = {
      onChange: jest.fn(),
    }
  })

  it('Should render FileUploader component', () => {
    component = mount(<FileUploader {...props} />)

    expect(component).toMatchSnapshot('FileUploader')
    expect(component.find(ELEMENT_SELECTORS.input).length).toBe(1)
  })

  it('Should render disabled FileUploader component with', () => {
    component = mount(<FileUploader {...props} disabled={true} />)

    expect(component).toMatchSnapshot('FileUploaderDisabled')
    expect(component.find(ELEMENT_SELECTORS.input).prop('disabled')).toBe(true)
  })

  it('Should call onChange', () => {
    component = mount(<FileUploader {...props} disabled={true} />)
    component
      .find(ELEMENT_SELECTORS.input)
      .simulate('change', {target: {files: [new File([], 'mock')]}})

    expect(props.onChange).toBeCalled()
  })
})
