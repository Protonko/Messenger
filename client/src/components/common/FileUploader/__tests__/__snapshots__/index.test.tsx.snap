// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileUploader Should render FileUploader component: FileUploader 1`] = `
<FileUploader
  onChange={[MockFunction]}
>
  <div
    className="file-uploader"
  >
    <label
      className="file-uploader__label"
    >
      <div
        className="file-uploader__icon"
      />
      <input
        className="file-uploader__input"
        name="file"
        onChange={[Function]}
        type="file"
      />
    </label>
  </div>
</FileUploader>
`;

exports[`FileUploader Should render disabled FileUploader component with: FileUploaderDisabled 1`] = `
<FileUploader
  disabled={true}
  onChange={[MockFunction]}
>
  <div
    className="file-uploader file-uploader--disabled"
  >
    <label
      className="file-uploader__label"
    >
      <div
        className="file-uploader__icon"
      />
      <input
        className="file-uploader__input"
        disabled={true}
        name="file"
        onChange={[Function]}
        type="file"
      />
    </label>
  </div>
</FileUploader>
`;
